File: infra/docker/Dockerfile.java
# Multi-stage build for Java services
FROM maven:3.9.6-eclipse-temurin-21-alpine AS builder

ARG SERVICE_NAME

WORKDIR /src
COPY services/${SERVICE_NAME}/pom.xml .
RUN mvn dependency:go-offline -B

COPY services/${SERVICE_NAME}/src ./src
RUN mvn clean package -DskipTests

# Runtime stage
FROM eclipse-temurin:21-jre-alpine AS runtime

ARG SERVICE_NAME

WORKDIR /app
COPY --from=builder /src/target/*.jar app.jar

# Run as non-root
RUN addgroup -S nonroot && adduser -S nonroot -G nonroot
USER nonroot:nonroot

ENTRYPOINT ["java", "-jar", "app.jar"]

File: infra/docker/Dockerfile.go
# Multi-stage build for Go services
FROM golang:1.23-alpine AS builder

ARG SERVICE_NAME

WORKDIR /src
# Use workspace/monorepo context
COPY . .
WORKDIR /src/services/${SERVICE_NAME}

RUN go mod download
# Build a static binary with no debug info
RUN CGO_ENABLED=0 GOOS=linux go build -ldflags="-s -w" -o /app/service .

# Use distroless for runtime
FROM gcr.io/distroless/static-debian12:nonroot AS runtime
COPY --from=builder /app/service /app/service
USER nonroot:nonroot
ENTRYPOINT ["/app/service"]
